src/System.Windows.Forms.Design/src/System/ComponentModel/Design/UndoEngine.UndoUnit.cs

```diff
@@ -0,0 +1,793 @@
+        ///  This undo event handles addition and removal of components.
+        /// </summary>
+        private sealed class AddRemoveUndoEvent : UndoEvent
+        {
+            private readonly SerializationStore _serializedData;
+            private readonly string? _componentName;
+
+            /// <summary>
+            ///  Creates a new object that contains the state of the event.  The last parameter, add, determines the initial mode of this event.  If true, it means this event is being created in response to a component add.  If false, it is being created in response to   a component remove.
+            /// </summary>
+            public AddRemoveUndoEvent(UndoEngine engine, IComponent component, bool add)
+            {
+                _componentName = component.Site!.Name;
+                NextUndoAdds = !add;
+                OpenComponent = component;
+
+                Debug.WriteLineIf(s_traceUndo.TraceVerbose, $"UndoEngine: ---> Creating {(add ? "Add" : "Remove")} undo event for '{_componentName}'");
+                using (_serializedData = engine._serializationService.CreateStore())
+                {
+                    engine._serializationService.Serialize(_serializedData, component);
+                }
+
+                // For add events, we commit as soon as we receive the event.
+                Committed = add;
+            }
+
+            /// <summary>
+            ///  Returns true if the add remove event has been comitted.
+            /// </summary>
+            internal bool Committed { get; private set; }
+
+            /// <summary>
+            ///  If this add/remove event is still open, OpenComponent will contain the component it is operating on.
+            /// </summary>
+            internal IComponent OpenComponent { get; }
+
+            /// <summary>
+            ///  Returns true if undoing this event will add a component.
+            /// </summary>
+            internal bool NextUndoAdds { get; private set; }
+
+            /// <summary>
+            ///  Commits this event.
+            /// </summary>
+            internal void Commit(UndoEngine engine)
+            {
+                if (!Committed)
+                {
+                    Debug.WriteLineIf(s_traceUndo.TraceVerbose, $"UndoEngine: ---> Committing remove of '{_componentName}'");
+                    Committed = true;
+                }
+            }
+
+            /// <summary>
+            ///  Actually performs the undo action.
+            /// </summary>
+            public override void Undo(UndoEngine engine)
+            {
+                if (NextUndoAdds)
+                {
+                    Debug.WriteLineIf(s_traceUndo.TraceVerbose, $"UndoEngine: ---> Adding '{_componentName}'");
+                    // We need to add this component.  To add it, we deserialize it and then we add it to the designer host's container.
+                    if (engine.GetRequiredService(typeof(IDesignerHost)) is IDesignerHost host)
```

Comment:
should this be ? or even move out of the `if-else` block?
```suggestion
                    IDesignerHost host = engine.GetRequiredService<IDesignerHost>();
```

Reply:
I didn't make this change because the generic version of `GetRequiredService` is slighly different than the non-generic one:

- `GetRequiredService<T>` throws if we don't find a service of type T
- `GetRequiredService` throws only if `GetService` returns null. It's fine with it returning an object of a different type than the one requested. I have no idea if it's actually possible though (it's certainly not the contract of `GetService`)

I figured it would be a behaviour change here to use the generic version (even though I'd very much like to!), so I chose to live this as is

Reply:
Though the difference, i do not expect it to be the case.  Happy to go with Net9.0 as well but simplifying the code is an improvement here.

Reply:
Okay, I made the change